lib/Push.php

```diff
@@ -154,21 +202,38 @@ public function pushToDevice(int $id, INotification $notification, ?OutputInterf
 			return;
 		}
 
+		if ($this->deferPreparing) {
+			$this->notificationsToPush[$id] = clone $notification;
+			$this->loadDevicesForUsers[] = $notification->getUser();
+			$this->loadStatusForUsers[] = $notification->getUser();
+			return;
+		}
+
 		$user = $this->createFakeUserObject($notification->getUser());
 
-		$userStatus = $this->userStatusManager->getUserStatuses([
-			$notification->getUser(),
-		]);
+		if (!array_key_exists($notification->getUser(), $this->userStatuses)) {
+			$userStatus = $this->userStatusManager->getUserStatuses([
+				$notification->getUser(),
+			]);
 
-		if (isset($userStatus[$notification->getUser()])) {
-			$userStatus = $userStatus[$notification->getUser()];
+			$this->userStatuses[$notification->getUser()] = $userStatus[$notification->getUser()] ?? null;
+		}
+
+		if (isset($this->userStatuses[$notification->getUser()])) {
+			$userStatus = $this->userStatuses[$notification->getUser()];
 			if ($userStatus->getStatus() === IUserStatus::DND) {
 				$this->printInfo('<error>User status is set to DND - no push notifications will be sent</error>');
 				return;
 			}
 		}
 
-		$devices = $this->getDevicesForUser($notification->getUser());
+		if (!array_key_exists($notification->getUser(), $this->userDevices)) {
+			$devices = $this->getDevicesForUser($notification->getUser());
+			$this->userDevices[$notification->getUser()] = $devices;
+		} else {
+			$devices = $this->userDevices[$notification->getUser()] ?? [];
```

Comment:
It should not be necessary to explicitly set an empty array, as `getDevicesForUser()` should return it if there are no devices.

Even if it is set to be on the safe side it would be better to set it in `$this->userDevices` instead for consistency with how it is done in `$this->userStatuses`

Reply:
No, this is from an earlier approach where instead of `array_key_exists` I had a variable indicating if devices would be loaded already. Since we now do the empty array setting in `flushPayloads()` we can skip this, yes.